topology/probes/libvirt/libvirt.go

```diff
@@ -0,0 +1,309 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package libvirt
+
+import (
+	"context"
+	"encoding/xml"
+	"fmt"
+	"sync"
+
+	"github.com/skydive-project/skydive/config"
+
+	libvirtgo "github.com/libvirt/libvirt-go"
+	"github.com/skydive-project/skydive/logging"
+	"github.com/skydive-project/skydive/topology/graph"
+)
+
+// Probe is the libvirt probe
+type Probe struct {
+	sync.Mutex
+	graph.DefaultGraphListener                       // Probe is a listner
+	graph                      *graph.Graph          // the graph
+	conn                       *libvirtgo.Connect    // libvirt conection
+	interfaceMap               map[string]*Interface // Found interfaces not yet connected.
+	cidLifecycle               int                   // libvirt callback id of monitor to unregister
+	cidDevAdded                int                   // second monitor on devices added to domains
+	uri                        string                // uri of the libvirt connection
+	cancel                     context.CancelFunc    // cancel function
+	root                       *graph.Node           // root node for ownership
+	tunProcessor               *graph.Processor      // metadata indexer
+}
+
+// Address describes the XML coding of the pci addres of an interface in libvirt
+type Address struct {
+	Type     string `xml:"type,attr"`
+	Domain   string `xml:"domain,attr"`
+	Bus      string `xml:"bus,attr"`
+	Slot     string `xml:"slot,attr"`
+	Function string `xml:"function,attr"`
+}
+
+// DomainStateMap stringifies the state of a domain
+var DomainStateMap map[libvirtgo.DomainState]string = map[libvirtgo.DomainState]string{
+	libvirtgo.DOMAIN_NOSTATE:     "Undefined",
+	libvirtgo.DOMAIN_RUNNING:     "UP",
+	libvirtgo.DOMAIN_BLOCKED:     "BLOCKED",
+	libvirtgo.DOMAIN_PAUSED:      "PAUSED",
+	libvirtgo.DOMAIN_SHUTDOWN:    "DOWN",
+	libvirtgo.DOMAIN_CRASHED:     "CRASHED",
+	libvirtgo.DOMAIN_PMSUSPENDED: "PMSUSPENDED",
+	libvirtgo.DOMAIN_SHUTOFF:     "DOWN",
+}
+
+// Interface is XML coding of an interface in libvirt
+type Interface struct {
+	Mac struct {
+		Address string `xml:"address,attr"`
+	} `xml:"mac"`
+	Target struct {
+		Device string `xml:"dev,attr"`
+	} `xml:"target"`
+	Address Address `xml:"address"`
+	Alias   struct {
+		Name string `xml:"name,attr"`
+	} `xml:"alias"`
+	Host *graph.Node `xml:"-"`
+}
+
+// Domain is the subset of XML coding of a domain in libvirt
+type Domain struct {
+	Interfaces []Interface `xml:"devices>interface"`
+}
+
+// queryDomain uses libvirt to get information on the interfaces of a domain.
+// domain is the libvirt domain queried and bag is an accumulator for results
+func (probe *Probe) queryDomain(
+	domain *libvirtgo.Domain, // domain to query
+	domainNode *graph.Node, // Node representing the domain
+	bag *[]*Interface, // to accumulate results
```

Comment:
nit: change the function to `func (probe *Probe) getDomainInterfaces(...) []*Interface`

Reply:
yes. No need for 2 phase and a bag to collect all itfs.